lib/Dialect/TTL/Transforms/ConvertTTLToTTKernel.cpp

```
@@ -0,0 +1,434 @@
+          rewriter.create<UnrealizedConversionCastOp>(op.getLoc(), targetTy, v);
+      return cast.getResult(0);
+    };
+
+    FailureOr<Value> convertedSrc = convertOperand(adaptor.getSrc());
+    if (failed(convertedSrc)) {
+      return rewriter.notifyMatchFailure(op, "failed to convert src type");
+    }
+    FailureOr<Value> convertedDst = convertOperand(adaptor.getDst());
+    if (failed(convertedDst)) {
+      return rewriter.notifyMatchFailure(op, "failed to convert dst type");
+    }
+
+    auto srcKind = classifySrc(*convertedSrc);
+    auto dstKind = classifyDst(*convertedDst);
+
+    // Tensor accessor -> CB
+    if (srcKind == CopySourceKind::TensorAccessor &&
+        dstKind == CopyDestKind::CircularBuffer) {
+      return lowerTensorToCB(op, *convertedSrc, rewriter, *typeConverter);
+    }
+
+    // CB -> Tensor accessor
+    if (srcKind == CopySourceKind::CircularBuffer &&
+        dstKind == CopyDestKind::TensorAccessor) {
+      return lowerCBToTensor(op, *convertedDst, rewriter, *typeConverter);
+    }
+
+    // Unsupported pairs for now.
+    return rewriter.notifyMatchFailure(op, [&](Diagnostic &diag) {
+      diag << "unsupported ttl.copy src/dst combination: src="
+           << op.getSrc().getType() << " dst=" << op.getDst().getType();
+    });
+  }
+};
+
+struct WaitLowering : OpConversionPattern<WaitOp> {
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(WaitOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    // TODO(ttl): Lower ttl.wait to TRID-specific barriers keyed by the transfer
+    // handle (read vs write barrier based on transfer direction). Issue: #87.
+    //
+    // MVP behavior: determine direction from the transfer handle type and emit
+    // the corresponding TTKernel global barrier. If the handle is untyped,
+    // conservatively emit both barriers.
+    if (auto kind = getTransferKindFromHandleType(adaptor.getXf().getType())) {
```

Looking at the tests, it seems like we are always failing into the else path. I think we should error in the else path rather than trying to do something wrong in the silent default case (and throughout, let's prioritize failing loudly for invalid cases rather than trying to emit best effort IR that we know is wrong). 

Maybe this can be an mlir verifier + assert?